drivers: input: CST816S: Add power management / motion interrupt support

Kampi opened the pull request:
- Add power management support
- Add Gesture / Motion interrupt

Kampi committed:
drivers: input: cst816s: Add power management

Add power management support

Closes: #100300

Signed-off-by: Daniel Kampert <DanielKampert@kampis-elektroecke.de>

JarmouniA reviewed (commented):
Although the driver is used by ZSWatch, that info is not relevant and should not be mentioned in the commit message.

fabiobaltieri commented on drivers/input/input_cst816s.c:
```diff
@@ -106,42 +126,57 @@ struct cst816s_data {
+	uint8_t points;
+	uint16_t x;
+	uint16_t y;
+};
```
not pm

fabiobaltieri reviewed (commented):
hey, sorry don't want to sound pedantic but the PM comment has a lot of stuff and not a lot of PM and is really hard to follow, this needs some more splits, did a pass anyway

faxe1008 commented on drivers/input/input_cst816s.c:
```diff
@@ -292,14 +327,104 @@ static int cst816s_init(const struct device *dev)
+	if (ret) {
+		LOG_WRN("Write lp_scan_i_dac 0x%02x failed (%d)", p->lp_scan_i_dac, ret);
+		return ret;
+	}
```
#define CST816S_REG_LP_AUTO_WAKEUP_TIME 0xF4
#define CST816S_REG_LP_SCAN_TH          0xF5
#define CST816S_REG_LP_SCAN_WIN         0xF6
#define CST816S_REG_LP_SCAN_FREQ        0xF7
#define CST816S_REG_LP_SCAN_I_DAC       0xF8
#define CST816S_REG_AUTOSLEEP_TIME      0xF9

Since they are continous can you maybe reorder the struct fields and write the entire struct at once?

faxe1008 commented:
You can use BUILD_ASSERT in the device dt macro to check them for valid range with IN_RANGE

Kampi commented on drivers/input/input_cst816s.c:
```diff
@@ -292,16 +320,87 @@ static int cst816s_init(const struct device *dev)
+                   .scan_win = DT_INST_PROP_OR(index, scan_win, 0),   \
+                   .scan_freq = DT_INST_PROP_OR(index, scan_freq, 0), \
+                   .scan_i_dac = DT_INST_PROP_OR(index, scan_i_dac, 0), \
+               },)) }; \
```
No 4 spaces but there were tabs mixed in. Copilot marked it already.

JarmouniA commented:
> So if I have implemented it, I can close it.

Only the reviewer can determine if it was actually implemented correctly, unless the requested change is fairly trivial and unambiguous.

If the author resolves my comments, as a reviewer I have to click on "show resolved" for every one to see what was the comment (I could have reviewed the PR 1 month ago) & to see the author's response if any, which is a lot of work.

Also, from my experience some authors mark comments they do not want to implement or discuss as resolved to hide them, which is unfortunate (not saying it's the case here).

fabiobaltieri commented:
Anyway we are steering this issue, @JarmouniA if you want to change the guideline I'd say just open a PR to change it and we can discuss that on the PR. I don't think it's that important, at the end of the day you can't rely on people following the guideline and have to double check anyway.

JarmouniA commented:
> Anyway we are steering this issue, @JarmouniA if you want to change the guideline

I'm fine with it as is.

fabiobaltieri commented on drivers/input/input_cst816s.c:
```diff
@@ -292,16 +320,123 @@ static int cst816s_init(const struct device *dev)
+	case PM_DEVICE_ACTION_TURN_ON:
+		ret = cst816s_apply_profile(cfg);
+		if (ret < 0) {
+			LOG_WRN("Could not apply suspend profile! (%d)", ret);
```
```suggestion
		if (ret < 0) {
			LOG_WRN("Could not apply suspend profile (%d)", ret);
```

Sorry I know this may be the most nitpicky comment ever, it's odd to have the extra emphasis on this, does not strike me as good use of flash memory :-)

Kampi commented on drivers/input/input_cst816s.c:
```diff
@@ -292,16 +320,123 @@ static int cst816s_init(const struct device *dev)
+#define CST816S_DEFINE(index)             \
+	IF_ENABLED(CONFIG_PM_DEVICE, (                                         \
+		BUILD_ASSERT(DT_INST_PROP(index, auto_wake_time) >= 1 &&       \
+				DT_INST_PROP(index, auto_wake_time) <= 5,       \
```
Haha nice. I've asked Claude how I can combine these macros and he suggested this and your combined approach. My approach was mentioned as the most usual one :D

But sure, I can combine them :)

Kampi commented on dts/bindings/input/hynitron,cst816s.yaml:
```diff
@@ -21,3 +21,49 @@ properties:
+  auto-wake-time:
+    type: int
+    default: 5
+    description:
```
I´ve thought about this solution yesterday, but I don´t like it. The reason is that a valid time is in range 1 to 5. I can use an enum from 1 to 5, for sure, but then I have to increase the index by 1 (or is there a better solution to let the enum start with index 1?)
That´s the reason why I decided to use a BUILD_ASSERT and check for a valid time there.

joelguittet replied:
My opinion on this is that a duration in general is not an enum so I'm fine with the original proposal

JarmouniA replied:
> I can use an enum from 1 to 5, for sure, but then I have to increase the index by 1

I don't understand what's the problem here.

JarmouniA replied:
> My opinion on this is that a duration in general is not an enum

!?? The enum is just a way for DT compiler to validate the values, has nothing to do with the type of the property.
https://docs.zephyrproject.org/latest/build/dts/bindings-syntax.html#enum

joelguittet replied:
Agree with you. But it's not only a technical DT point of view. We speak of documentation and user friendly presentation.
If i see an enum then i would expect an enum in the datasheet, but it's not.

Kampi replied:
> > I can use an enum from 1 to 5, for sure, but then I have to increase the index by 1
> 
> I don't understand what's the problem here.

There is no problem with your solution. If there is a general way to handle duration I will use it for sure. From my point of view it´s a bit more complicated to use this selection based approach and increase the index by one instead to use an assert to catch invalid values.

But as @joelguittet mentioned: if there is a general way of handling limited values for durations I will use this way to go.

Kampi replied:
@JarmouniA I´ve changed it to an enum. Please take a look at it.

joelguittet commented on drivers/input/input_cst816s.c:
```diff
@@ -233,14 +254,21 @@ static int cst816s_chip_init(const struct device *dev)
 		return -ENODEV;
 	}
 
+	ret = i2c_reg_update_byte_dt(&cfg->i2c, CST816S_REG_MOTION_MASK, CST816S_MOTION_EN_DCLICK,
```
I guess this not linked directly to PM (it's not conditioned on CONFIG_PM) so I would expect a separated commit for this modification, to explain the purpose

Kampi committed:
drivers: input: cst816s: Add motion interrupt

Add support for gesture / motion based interrupts
when INPUT_CST816S_EV_DEVICE is enabled.

Signed-off-by: Daniel Kampert <DanielKampert@kampis-elektroecke.de>

uLipe commented on drivers/input/input_cst816s.c:
```diff
@@ -88,6 +93,18 @@ struct cst816s_config {
 #endif
+
+#ifdef CONFIG_PM_DEVICE
+	/** cst816s power management profile. */
```
Nit: it would be good to document also the struct fields.

Kampi replied:
Will do

uLipe reviewed (commented):
Looks good, I just have a nit.

Kampi committed:
drivers: input: cst816s: Refactor errors and warnings

Add error code to error and warning log messages to
make debugging easier.

Signed-off-by: Daniel Kampert <DanielKampert@kampis-elektroecke.de>

fabiobaltieri commented on drivers/input/input_cst816s.c:
```diff
@@ -292,16 +315,108 @@ static int cst816s_init(const struct device *dev)
+                                                                                           \
+	DEVICE_DT_INST_DEFINE(index, cst816s_init, PM_DEVICE_DT_INST_GET(index), \
+			      &cst816s_data_##index, &cst816s_config_##index, POST_KERNEL, \
+			      CONFIG_INPUT_INIT_PRIORITY, NULL);
```
yeah that's good but then you have to align them by hand
